### ISSUE TITLE
awskinesisexporter: Add support for KPL aggregation

### ISSUE THREAD

**Issue by MovieStoreGuy** [open-telemetry/opentelemetry-collector-contrib#6094]
**Is your feature request related to a problem? Please describe.**
With the most recent updates into the exporter, KPL aggregation was one thing that was dropped and I would like to see come back.

**Describe the solution you'd like**
Can implement this: https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-concepts.html#kinesis-kpl-concepts-aggretation

**Describe alternatives you've considered**
The current method limits how much data most can be sent however, it can be optimised to include more per batch using the PutRecords api.

So instead of using splitting an incoming resource slice with one resource per batch, the exporter should be able to add multiple resources per record.

**Comment by mottibec** [comment]
Is anyone working on this? If not, I would like to.
I was thinking about using an existing lib for the KPL aggregation. It looks like aws don't have an official client to support this in go, but I ran into https://github.com/a8m/kinesis-producer.

**Comment by jpkrohling** [comment]
It's yours, @mottibec. Not sure about the library: if really necessary, we can certainly consider it, but it will make the dep tree more complex.

cc @anuraaga

**Comment by MovieStoreGuy** [comment]
Hey @mottibec,

I had purposely avoided using that library for a few reasons:
- The error handling path was problematic since the async sender would make dropped data harder to replay / retry 
- The Partition key is the same for an entire batch, so the chances of successful writes became problematic on smaller stream sizes but with high batching sizes.
- Not entirely sure it supported compression

We had tried it for a period of time internally at Atlassian and found the max throughput we could achieve was a batch size of 1 and even then it was problematic due to how large payloads could be following the semantic convention for traces and an internally defined one for metrics.

Once we moved over to the current exporter, we found the number of successful writes increase on smaller streams since each message had a unique (not the same) partition key and retries, timeouts, and queueing could all be implemented by the existing export helpers within the project.

**Comment by mottibec** [comment]
@MovieStoreGuy So, if I understand you correctly, you suggest not using KPL aggregation at all and sending batch requests with the current exporter? 

The primary motivation for us at aspecto to use the KPL is for computability with other components in our pipeline. At the moment, we are stuck on an old version of the collector with the kinesis exporter that supports KPL aggregation. But if it won't increase our throughput because of the batch size, it might be worth refactoring the other components in our pipeline instead and using the exporter as-is.

**Comment by MovieStoreGuy** [comment]
> @MovieStoreGuy So, if I understand you correctly, you suggest not using KPL aggregation at all and sending batch requests with the current exporter?

Not exactly, more saying that the reference library is something that we at Atlassian spiked and found problematic hence making the custom client in the collector today. 

KPL aggregation can be completely possible, just requires a bit of effort to achieve since the exporter will attempt to write the smallest atomic representation of the telemetry per given record; it can be done, and I encourage doing so, just yet to find the time to update the component to do so (partially time management, partially allowing others on my team to contribute to the project) or drop in lib that works well with the processing model used in the collector.

**Comment by aallawala** [comment]
Still relevant and I'd like to help. Two things worth adding:

**This won't arrive from the SDK.** aws-sdk-go-v2 has no aggregation support and no kinesis `feature/` module. `awslabs/kinesis-aggregation` ships "Aggregation & Deaggregation" for Java, Node and Python, but the Go module is titled ["Go Kinesis **Deaggregation** Modules"](https://github.com/awslabs/kinesis-aggregation/tree/master/go) — deaggregator only. When AWS needed a Go aggregator they vendored one: [`amazon-kinesis-streams-for-fluent-bit/aggregate/`](https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/tree/mainline/aggregate), 207 lines, Apache-2.0. So this is component-side work either way, but with a license-compatible reference — which also sidesteps @jpkrohling's dep-tree concern about `a8m/kinesis-producer`.

**@MovieStoreGuy — two of your 2022 objections look answerable from that reference.** On partition keys: AWS's mitigation isn't shard-mapping, it's keeping aggregates small — [`defaultMaxAggRecordSize = 20 * 1024`](https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/cfe06d78e31002ed2b2c977034cf283e02ec55a3/aggregate/aggregator.go#L23), not the 1 MB ceiling, with [`pkeys[0]`](https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/cfe06d78e31002ed2b2c977034cf283e02ec55a3/aggregate/aggregator.go#L157) as the outer key. At 20 KB this exporter would still emit many randomly-keyed records per flush, so the shard spread gained in #5248 survives at aggregate granularity — I suspect the "batch size of 1" result came from aggregating toward 1 MB. On compression: the [magic number](https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/cfe06d78e31002ed2b2c977034cf283e02ec55a3/aggregate/aggregator.go#L17) must be the first four bytes, so the aggregate itself can never be compressed — compression stays per-record inside it, where `AddRecord` already does it. Your error-handling objection stands, though: a rejected aggregate takes every record in it down, which is an independent argument for that small default.

**On #8114** — it wasn't rejected; the stale bot closed it and the reopen never happened. The only thing left genuinely unresolved was your question of whether aggregation belongs in `Batch`, instead of it, or as a step after marshal. I'd argue a discrete step between marshal and batch behind an `Aggregator` interface with a no-op impl (close to the `NoOpAggregator` you suggested) — keeps `Batch` unaware of aggregation, keeps `Chunk()` side-effect-free, and avoids the `Close()`-style lifecycle you flagged as a smell.

If that matches your thinking I'm happy to pick this up. @mottibec — you're still assigned; still interested? Happy to revive #8114 so your work keeps the attribution, or open a fresh PR crediting you.
